csharp/client/Dhe_NetClient/session/SessionManager.cs

```diff
@@ -0,0 +1,271 @@
+      var result = dict.Values.FirstOrDefault(i => i.Config.Name == pqName);
+      if (result == null) {
+        throw new Exception($"pq name='{pqName}' not found");
+      }
+      return result;
+    });
+    return DndClient.Create(pqSerial, this, client, removeOnClose);
+  }
+
+  public DndClient ConnectToPqById(Int64 pqSerial, bool removeOnClose) {
+    var (_, client) = FindPqAndConnect(dict => {
+      if (!dict.TryGetValue(pqSerial, out var result)) {
+        throw new Exception($"pqSerial='{pqSerial}' not found");
+      }
+      return result;
+    });
+    return DndClient.Create(pqSerial, this, client, removeOnClose);
+  }
+
+  private (Int64, Client) FindPqAndConnect(
+    Func<IReadOnlyDictionary<Int64, PersistentQueryInfoMessage>, PersistentQueryInfoMessage> filter) {
+    using var subscription = _controllerClient.Subscribe();
+    if (!subscription.Current(out var version, out var configMap)) {
+      throw new Exception("Controller subscription has closed");
+    }
+
+    var info = filter(configMap);
+    var pqSerial = info.Config.Serial;
+    var pqName = info.Config.Name;
+
+    while (true) {
+      // It may make sense to have the ability to provide a timeout
+      // in the future; absent a terminal or running state being found or
+      // the subscription closing, as is this will happily run forever.
+      var status = info.State.Status;
+      if (ControllerClient.IsTerminal(status)) {
+        throw new Exception($"pqName='{pqName}', pqSerial={pqSerial} " +
+          $"is in terminal state={info.State.Status}");
+      }
+
+      if (ControllerClient.IsRunning(status)) {
+        break;
+      }
+
+      if (!subscription.Next(version) ||
+          !subscription.Current(out version, out configMap)) {
+        throw new Exception("Controller subscription has closed");
+      }
+
+      if (!configMap.TryGetValue(pqSerial, out info)) {
+        throw new Exception($"pqName='{pqName}', pqSerial={pqSerial} " +
+          $"is no longer available");
+      }
+    }
+
+    return ConnectToPq(info);
+  }
+
+  private (Int64, Client) ConnectToPq(PersistentQueryInfoMessage infoMsg) {
+    var url = infoMsg.State.ConnectionDetails.GrpcUrl;
+    var pqSerial = infoMsg.Config.Serial;
+
+    var pqStrForErr = $"pq_name='{infoMsg.Config.Name}', pq_serial={pqSerial}";
+    var pos = url.IndexOf(':');
+    const int targetOffsetAfterColon = 3;  // "://"
+    if (url.IsEmpty()) {
+      if (infoMsg.State.EngineVersion.IsEmpty()) {
+        throw new Exception($"{pqStrForErr} is not a Community engine");
+      }
+      throw new Exception($"{pqStrForErr} has no gRPC connectivity available");
+    }
+
+    // pos < 1 means 0 or not found
+    if (pos < 1 || pos + targetOffsetAfterColon >= url.Length) {
+      throw new Exception($"{pqStrForErr} has invalid url='{url}'");
+    }
+
+    var urlSchema = url.Substring(0, pos);
+    var urlTarget = url.Substring(pos + targetOffsetAfterColon);

```

Comment:
This clearly works, but aren't there standard libraries for parsing URL/URI rather than another hand-rolled version?